### PR TITLE
eve: async mode for redis output - v7

### PR DESCRIPTION
**Pull request by fooinha**
doc: async: true ## if redis replies are read asynchronously
eve: detects libevent for async redis at configure
eve: moves redis output code to new file - util-log-redis.{c,h}
eve: redis pipeline batch commands with async mode
eve: redis ECHO and QUIT commands for async mode
eve: redis output defaults if conf is missing

Replaces #2576 

@jasonish

### CODE REVIEW

File: src/output.c [hunk 1]
```diff
@@ -1002,8 +1006,16 @@ void OutputRegisterRootLogger(ThreadInitFunc ThreadInit,
     TAILQ_INSERT_TAIL(&RootLoggers, logger, entries);
 }
 
+//XXX FIXME  - What the best place and names for this
+void OutputLibEventUsePthreads() {
+#ifdef HAVE_LIBEVENT_PTHREADS
+    evthread_use_pthreads();
+#endif
+}
+
```
Comment (jasonish) [inline src/output.c]: Perhaps do this in util-log-redis.c, SCLogRedisInit(), then you can do the ifdef in there, and keep the libevent stuff out of src/output.c as libevent usage is only for redis, for now.

File: src/util-log-redis.c [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+ *         Emits and awaits response for an async QUIT command.
+ *         It's used to disconnect with redis
+ *  \param ctx redis context
+ */
+static void SCLogAsyncRedisSendQuit(SCLogRedisContext * ctx)
+{
+    if (ctx->connected) {
+        redisAsyncCommand(ctx->async, SCRedisAsyncQuitCommandCallback, ctx, "QUIT");
+        SCLogInfo("QUIT Command sent to redis. Connection will terminate!");
+        event_base_dispatch(ctx->ev_base);
```
Comment (jasonish) [inline src/util-log-redis.c]: I'm not sure about the use of event_base_dispatch here.  It basically makes this call blocking, as its not going to return until event_base_breakloop() is called.  Is there a reason these can't be event_base_loop(ctx->ev_base, EVLOOP_NONBLOCK) like used in the Write message which are non-blocking from the perspective of Suricata?

**Comment by jasonish** [comment]
Overall this is working pretty well.  ASAN is happy.  Connect/reconnect work well.

Async has the advantage of not blocking Suricata during write if the socket becomes backed up, which could happen with the non-async mode, just like unix_stream.

I'd still like to run it for an extended period of time though.

File: src/util-log-redis.c [hunk 3]
```diff
@@ -0,0 +1,602 @@
...
+static int SCConfLogReopenAsyncRedis(LogFileCtx *log_ctx)
+{
+    SCLogRedisContext * ctx = log_ctx->redis;
+    const char *redis_server = log_ctx->redis_setup.server;
+    int redis_port = log_ctx->redis_setup.port;
+
+    /* only try to reconnect once per second */
+    if (log_ctx->redis_setup.tried >= time(NULL)) {
+        return -1;
+    }
```
Comment (jasonish) [inline src/util-log-redis.c]: This never seems to be true, so it attempts to reconnect many times per second, at least with auto-fp.  tried is reset to 0 at the bottom of this function probably causing this.

Comment (fooinha) [inline src/util-log-redis.c]: I'll try to fix this and keep behavior ...

File: doc/userguide/configuration/suricata-yaml.rst [hunk 4]
```diff
@@ -267,6 +267,7 @@ integration with 3rd party tools like logstash.
       #redis:
       #  server: 127.0.0.1
       #  port: 6379
+      #  async: true ## if redis replies are read asynchronously
```
Comment (jasonish) [inline doc/userguide/configuration/suricata-yaml.rst]: I thought this was on the wiki but its not, documentation commits should be in their own commit please, separate from code commits.

Comment (fooinha) [inline doc/userguide/configuration/suricata-yaml.rst]: two different commits now.

File: src/output.c (hunk 1 above)
Comment (fooinha) [inline src/output.c]: moved this to util-log-redis.c ... and it's called only if needed by output json init register.

File: src/util-log-redis.c (hunk 2 above)
Comment (fooinha) [inline src/util-log-redis.c]: Yes! ... replaced this one to event_base_loop(ctx->ev_base, EVLOOP_NONBLOCK)

File: src/util-log-redis.c (hunk 3 above)
Comment (fooinha) [inline src/util-log-redis.c]: Fixed this, also the state variables that control this were moved from redis setup struct to context struct.

File: src/util-log-redis.c (hunk 2 above)
Comment (fooinha) [inline src/util-log-redis.c]: we'll keep  event_base_dispatch(ctx->ev_base) here , because we're terminating, otherwise we may leak... 
ASAN confirms.
I'll re-visit this.

**Comment by fooinha** [comment]
Replaced by #2604
